File: app/api/opml.py
from tempfile import TemporaryFile
from xml.etree import ElementTree as ET
from flask import send_file

from models import Category, Subscription
from api import api_bp


@api_bp.route("/opml/export/", methods=["GET"])
def opml_export():
    data = Subscription.query.outerjoin(Category).with_entities(
        Category.title, Subscription.title,
        Subscription.feed_url, Subscription.web_url).all()

    root = ET.Element("opml", attrib={"version": "1.0"})
    ET.SubElement(root, "head")
    body = ET.SubElement(root, "body")

    for category in set([i[0] for i in data]):
        node = body

        if category is not None:
            attrib = {"text": category, "title": category}
            node = ET.SubElement(body, "outline", attrib=attrib)

        for subscription in filter(lambda i: i[0] == category, data):
            attrib = {
                "type": "rss",
                "text": subscription[1],
                "title": subscription[1],
                "xmlUrl": subscription[2],
                "webUrl": subscription[3] or ""}
            ET.SubElement(node, "outline", attrib=attrib)

    file = TemporaryFile()
    ET.ElementTree(root).write(file, encoding="utf-8", xml_declaration=True)

    return send_file(
        file, mimetype="application/xml",
        as_attachment=True, attachment_filename="feeds.opml")

File: static/ts/util/opml.ts
import { OPMLItem, Category, Subscription } from "data/types";
import * as categories from "data/categories";
import * as subscriptions from "data/subscriptions";

const PROLOG = '<?xml version="1.0" encoding="UTF-8"?>';


export function parse(opml: string): OPMLItem[] {
	let parser = new DOMParser();
	let doc = parser.parseFromString(opml, "application/xml");
	let nodes = Array.from(doc.querySelectorAll("outline"))
		.filter(node => node.getAttribute("type") == "rss");

	return nodes.map(node => ({
		title: node.getAttribute("title"),
		xmlUrl: node.getAttribute("xmlUrl"),
		webUrl: node.getAttribute("webUrl"),
		category: node.closest("body > outline")?.getAttribute("title") || null
	}));
}

export function serialize() {
	let doc = document.implementation.createDocument(null, null, null);
	let body = doc.createElement("body");

	categories.list()
		.forEach(c => body.appendChild(buildCategory(doc, c)));

	subscriptions.list()
		.filter(s => !s.category_id)
		.forEach(s => body.appendChild(buildSubscription(doc, s)));

	let root = doc.createElement("opml");
	root.setAttribute("version", "1.0");
	doc.appendChild(root);

	root.appendChild(doc.createElement("head"));
	root.appendChild(body);

	return `${PROLOG}${(new XMLSerializer()).serializeToString(doc)}`;
}

function buildCategory(doc: XMLDocument, data: Category) {
	let node = doc.createElement("outline");
	node.setAttribute("text", data.title);
	node.setAttribute("title", data.title);
	subscriptions.list(data.id)
		.forEach(s => node.appendChild(buildSubscription(doc, s)));
	return node;
}

function buildSubscription(doc: XMLDocument, data: Subscription) {
	let node = doc.createElement("outline");
	node.setAttribute("type", "rss");
	node.setAttribute("text", data.title);
	node.setAttribute("title", data.title);
	node.setAttribute("title", data.title);
	node.setAttribute("xmlUrl", data.feed_url);
	data.web_url && node.setAttribute("webUrl", data.web_url);
	return node;
}
